network/tools/network-policy-enforcement-latency/pkg/client.go

```diff
@@ -0,0 +1,545 @@
+		c.config.workerCount = utils.DefaultWorkerCount
+		klog.Infof("workerCount value is less than 1. The workerCount value is set to the default (%d)", utils.DefaultWorkerCount)
+	} else if c.config.workerCount > utils.MaxWorkerCount {
+		c.config.workerCount = utils.MaxWorkerCount
+		klog.Infof("workerCount value is higher than the maximum allowed. The workerCount value is set to the maximum (%d)", utils.MaxWorkerCount)
+	}
+
+	testTypeName := "pod-creation-enforcement-latency"
+	if !c.config.testPodCreation {
+		if len(c.config.allowPolicyName) == 0 {
+			return fmt.Errorf("allowPolicyName is not specified for policy creation test")
+		}
+		testTypeName = "policy-creation-enforcement-latency"
+	}
+
+	klog.Infof("Connectivity test client mode: %q", testTypeName)
+
+	// Log all test client config flags.
+	klog.Infof("Network policy enforcement latency test client parameters:")
+	klog.Infof("--testClientNamespace=%s", c.config.testClientNamespace)
+	klog.Infof("--targetLabelSelector=%s", c.config.targetLabelSelector)
+	klog.Infof("--targetNamespace=%s", c.config.targetNamespace)
+	klog.Infof("--targetPort=%d", c.config.targetPort)
+	klog.Infof("--expectedTargets=%d", c.config.expectedTargets)
+	klog.Infof("--allowPolicyName=%s", c.config.allowPolicyName)
+	klog.Infof("--metricsPort=%d", c.config.metricsPort)
+	klog.Infof("--testPodCreation=%t", c.config.testPodCreation)
+	klog.Infof("--workerCount=%d", c.config.workerCount)
+
+	return nil
+}
+
+// measureNetPolicyCreation runs the network policy enforcement latency test for
+// network policy creation.
+func (c *TestClient) measureNetPolicyCreation() error {
+	klog.Infof("Starting to measure network policy enforcement latency on network policy creation")
+	podList, err := c.targetPodList()
+	if err != nil {
+		return fmt.Errorf("failed to get the pod list, error: %v", err)
+	}
+	klog.Infof("%d pods listed", len(podList))
+
+	c.policyCreatedTime = &timeWithLock{lock: sync.RWMutex{}}
+	wg := sync.WaitGroup{}
+
+	// Keep sending requests to all pods until all of them are reached.
+	for _, pod := range podList {
+		target := &targetSpec{
+			ip:        pod.Status.PodIP,
+			name:      pod.GetName(),
+			namespace: pod.GetNamespace(),
+		}
+
+		wg.Add(1)
+		go func() {
+			c.recordFirstSuccessfulRequest(target, false)
+			wg.Done()
+		}()
+	}
+
+	wg.Wait()
+	return nil
+}
+
+// targetPodList returns a list of pods based on the test client config, for
+// target namespace and target label selector, with a limit to the listed pods.
+func (c *TestClient) targetPodList() ([]corev1.Pod, error) {
+	podList, err := c.k8sClient.CoreV1().Pods(c.config.targetNamespace).List(context.Background(), metav1.ListOptions{LabelSelector: c.config.targetLabelSelector, Limit: int64(c.config.expectedTargets)})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list pods for selector %q in namespace %q, error: %v", c.config.targetLabelSelector, c.config.targetNamespace, err)
+	}
+
+	if len(podList.Items) == 0 {
```

Comment:
Shouldn't it be equal to expectedTargets?

Reply:
Changed the flag name to `maxTargets`, because that's what it actually does.
There will be an error if there are no target pods -- the test cannot carry on.
It will not fail if there are more than `maxTargets`, it will just limit the kube API list call to that value.

Reply:
what if there is just one pod returned?

Reply:
It will run the test for one pod. Send requests to it until it's reached, and generate one latency metric data point for it.